File: Strata/trunk/java/components/StrataInspector/src/strata1/inspector/diagnostic/DiagnosticSuite.java
// ##########################################################################
// # File Name:	DiagnosticSuite.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataInspector Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataInspector Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataInspector
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.inspector.diagnostic;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Represents a suite of {@code IDiagnostic}s that are meant to be 
 * run as a whole or as part of a larger suite. {@code DiagnosticSuite} 
 * is a <a href="{@docRoot}/designpatterns/pat4c.htm">Composite</a> of 
 * {@code IDiagnostic}s.
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class DiagnosticSuite 
	extends AbstractDiagnostic
{
	private Map<String,IDiagnostic> itsDiagnostics;
	
	/************************************************************************
	 * Creates a new DiagnosticSuite. 
	 *
	 * @param name	The name of the suite of diagnostics.
	 */
	public 
	DiagnosticSuite(String name)
	{
		super( name );
		itsDiagnostics = new LinkedHashMap<String,IDiagnostic>();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public void 
	runDiagnostic(IDiagnosticResult result)
	{
		try
		{
			result.beginDiagnostic( this );
			beginDiagnosticMode();
			
			for (IDiagnostic d:itsDiagnostics.values())
				d.runDiagnostic( result );
		}
		catch (DiagnosticAbortedException ae)
		{
			result.reportBeginFailure( this,ae );
		}
		finally
		{
			endDiagnosticMode();
			result.endDiagnostic( this );
		}
	}

	/************************************************************************
	 * Adds a IDiagnostic to the DiagnosticSuite. 
	 *
	 * @param 			d	The diagnostic being added.
	 * @precondition	this.hasDiagnostic( d.getName() ) == false
	 * @postcondition	this.hasDiagnostic( d.getName() ) == true
	 * @postcondition	this.getDiagnostic( d.getName() ) == d
	 */
	public void 
	addDiagnostic(IDiagnostic d)
	{
		itsDiagnostics.put( d.getName(),d );
	}
	
	/************************************************************************
	 * Removes the IDiagnostic with the specified 
	 * name from the DiagnosticSuite. 
	 *
	 * @param 			name	The name of the diagnostic being removed 
	 * 							from the suite.
	 * @precondition	this.hasDiagnostic( name ) == true
	 * @postcondition	this.hasDiagnostic( name ) == false
	 */
	public void 
	removeDiagnostic(String name)
	{
		itsDiagnostics.remove( name );
	}
	
	/************************************************************************
	 * Returns the IDiagnostic associated with the specified name. 
	 *
	 * @param 	name	The name of the diagnostic to be returned.
	 * @return	The diagnostic associated with name.
	 */
	public IDiagnostic 
	getDiagnostic(String name)
	{
		return itsDiagnostics.get( name );
	}
	
	/************************************************************************
	 * Queries the DiagnosticSuite if it contains a IDiagnostic 
	 * with the specified name. 
	 *
	 * @param 	name	The name of the diagnostic being queried.
	 * @return	True if the suite contains a diagnostic with the 
	 * 			specified name.
	 */
	public boolean 
	hasDiagnostic(String name)
	{
		return itsDiagnostics.containsKey( name );
	}
}

//##########################################################################

File: csharp/Components/Strata.Diagnostic.Core/Common/DiagnosticSuite.cs
//  ##########################################################################
//  # File Name: DiagnosticSuite.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Strata.Diagnostic.Core.Common
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// Represents a suite of <c>IDiagnostic</c>s that are meant to be 
    /// run as a whole or as part of a larger suite. <c>DiagnosticSuite</c> 
    /// is a <a href=
    /// "http://en.wikipedia.org/wiki/Composite_pattern">Composite</a> of 
    /// <c>IDiagnostic</c>s.
    /// </summary>
    ///  
    public 
    class DiagnosticSuite:
        AbstractDiagnostic
    {
        private readonly IDictionary<string,IDiagnostic> diagnostics;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>DiagnosticSuite</c> instance.
        /// </summary>
        /// 
        public 
        DiagnosticSuite(string name):
            base( name )
        {
            diagnostics = new Dictionary<string,IDiagnostic>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
	    public override async Task<IDiagnosticResult> 
        RunDiagnostic(IDiagnosticResult result)
	    {
		    try
		    {
			    result.BeginDiagnostic( this );
			    await BeginDiagnosticMode();
			
			    foreach (IDiagnostic d in diagnostics.Values)
				    await d.RunDiagnostic( result );
		    }
		    catch (DiagnosticAbortedException ae)
		    {
			    result.ReportBeginFailure( this,ae );
		    }
		    finally
		    {
			    await EndDiagnosticMode();
			    result.EndDiagnostic( this );
		    }

            return result;
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Adds a <c>IDiagnostic</c> to 
        /// the <c>DiagnosticSuite</c>.
        /// </summary>
        /// <param name="d">diagnostic being added to suite</param>
        /// <precondition>
        /// this.HasDiagnostic( d.Name ) == false
        /// </precondition>
        /// <postcondition>
        /// this.HasDiagnostic( d.Name ) == true
        /// </postcondition>
        /// <postcondition>
        /// this.GetDiagnostic( d.Name ) == d
        /// </postcondition>
        /// 
        public DiagnosticSuite
        AddDiagnostic(IDiagnostic d)
	    {
		    diagnostics.Add( d.Name,d );
            return this;
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Removes the <c>IDiagnostic</c> with the specified 
        /// name from the <c>DiagnosticSuite</c>.
        /// </summary>
        /// <param name="name">
        ///     Name of diagnostic begin removed from suite
        /// </param>
        /// <precondition>
        /// this.HasDiagnostic( name ) == true
        /// </precondition>
        /// <postcondition>
        /// this.HasDiagnostic( name ) == false
        /// </postcondition>
        /// 
        public DiagnosticSuite
        RemoveDiagnostic(string name)
	    {
		    diagnostics.Remove( name );
            return this;
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Returns the <c>IDiagnostic</c> associated 
        /// with the specified name.
        /// </summary>
        /// 
        /// <param name="name">Name of the diagnostic being queried</param>
        /// <returns>diagnostic associated with name</returns>
        /// 
	    public IDiagnostic 
	    GetDiagnostic(string name)
	    {
		    return diagnostics[name];
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Returns the <c>IDiagnostic</c> associated 
        /// with the specified name.
        /// </summary>
        /// 
        /// <param name="name">Name of the diagnostic being queried</param>
        /// <returns>diagnostic associated with name</returns>
        /// 
        public ICollection<IDiagnostic>
        GetDiagnostics()
        {
            return diagnostics.Values;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Queries the <c>DiagnosticSuite</c> if it contains an
        /// <c>IDiagnostic</c> with the specified name.
        /// </summary>
        /// 
        /// <param name="name">Name of the diagnostic being queried</param>
        /// <returns>
        /// True if the suite contains a diagnostic with the specified name.
        /// </returns>
        /// 
        public bool 
	    HasDiagnostic(string name)
	    {
		    return diagnostics.ContainsKey( name );
	    }

    }
}

//  ##########################################################################
